shell/common/rasterizer.cc

```diff
@@ -264,9 +267,31 @@ RasterStatus Rasterizer::DoDraw(
   // TODO(liyuqian): in Fuchsia, the rasterization doesn't finish when
   // Rasterizer::DoDraw finishes. Future work is needed to adapt the timestamp
   // for Fuchsia to capture SceneUpdateContext::ExecutePaintTasks.
-  timing.Set(FrameTiming::kRasterFinish, fml::TimePoint::Now());
+  const auto raster_finish_time = fml::TimePoint::Now();
+  timing.Set(FrameTiming::kRasterFinish, raster_finish_time);
   delegate_.OnFrameRasterized(timing);
 
+  const fml::TimePoint latest_frame_target_time =
+      delegate_.GetLatestFrameTargetTime();
+  if (latest_frame_target_time > frame_target_time) {
```

Comment:
latest_frame_target_time may not be updated as soon as the vsync happens which means there may be a millisecond or two before we consider this frame to have "lagged".

Reply:
The other place we could update it is in `vsync_waiter`. For example in iOS as soon as we get `onDisplayLink` (or right afterwards in `FireCallback` which will work across all platforms), we can update it this would avoid the `VsyncSchedulingOverhead`. I wasn't able to come up with a nice abstraction that would let us store this value there are consume it on the rasterizer.

Maybe we pass in a shared reference to a time recorder to both rasterizer and vsync waiter to record and consume these events. Thoughts?

Reply:
I think the time between when the vsync happens and when the first line of our code is called is the most of the delay. By the time we field the notification, it is a very tiny delta to when we update the delegate. So moving the update earlier only improves the issue a small amount.

I guess my point was to use the comparison `now > frame_target_time` instead. Note that a few lines before that, we already compute "now" anyway I believe.

If the new "next target time" hasn't been delivered in a callback yet then the value we record in the event won't be very useful as it will be an out of date value. We could get fancy and try to project forward when the next target time will be based on the prior target time. Something like

```
  if (now > frame_target_time) {
    ...
    TimePoint next_frame_target = latest_frame_target_time;
    if (now > next_frame_target) {
      next_frame_target += frame_budget_millis;
    }
    TraceEvent(frame_target_time, now, "latest target time", next_frame_target, ...);
  }
```

Reply:
Done!